File: Lesson34_PCF8591_LED/code/python/Lesson34_PCF8591_LED.py
import smbus
import time
from gpiozero import LED

bus = smbus.SMBus(1)
channels = [0x40, 0x41, 0x42, 0x43]

def setup(Addr):
    global address
    address = Addr

def read(chn):
    bus.write_byte(address, channels[chn])
    bus.read_byte(address)  # Dummy read for PCF8591
    return bus.read_byte(address)

def write(val):
    bus.write_byte_data(address, 0x40, int(val))

class Led:
    def __init__(self, pin):
        # BCM pin, common anode LED (active low)
        self.led = LED(pin, active_high=False)

    def on(self):
        """Turn on LED"""
        self.led.on()

    def off(self):
        """Turn off LED"""
        self.led.off()

    def destroy(self):
        """Close GPIO pin"""
        self.led.close()


if __name__ == '__main__':
    setup(0x48)
    light_threshold = 100
    # BCM17 (physical pin 11)
    led = Led(17)

    try:
        while True:
            light_intensity = read(0)
            print('Light intensity:', light_intensity)
            if light_intensity > light_threshold:
                led.on()
            else:
                led.off()
            time.sleep(1)
    except KeyboardInterrupt:
        led.destroy()

File: Lesson34_PCF8591_LED/code/c/Lesson34_PCF8591_LED.c
#include <stdio.h>
#include <lgpio.h>

#define PCF8591_ADDR  0x48
#define A0            0x40  // Photoresistor
#define LED_PIN       17    // BCM17 (wiringPi 0)

int main()
{
    int h = lgGpiochipOpen(4);
    if (h < 0) {
        printf("Failed to open gpiochip4, try open gpiochip0\n");
        h = lgGpiochipOpen(0);
        if (h < 0) return -1;
    }

    int fd = lgI2cOpen(1, PCF8591_ADDR, 0);
    if (fd < 0) {
        printf("Failed to open I2C device\n");
        return -1;
    }

    float light_intensity;
    int light_threshold = 100;

    lgGpioClaimOutput(h, 0, LED_PIN, 1);

    while (1) {
        light_intensity = lgI2cReadByteData(fd, A0);
        printf("Light intensity: %.2f\n", light_intensity);

        if (light_intensity > light_threshold)
            lgGpioWrite(h, LED_PIN, 0);  // LED on
        else
            lgGpioWrite(h, LED_PIN, 1);  // LED off

        lguSleep(0.5);
    }

    lgI2cClose(fd);
    lgGpiochipClose(h);
    return 0;
}
